File: apps/fcc-extension/internal/protocol/spend_window_test.go
package protocol

import (
	"encoding/json"
	"math/big"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"testing"
)

func TestSpendWindowTotalsV1CapsPrivateHistory(t *testing.T) {
	entries := make([]SpendWindowEntryV1, MaxSpendWindowEntriesV1+1)
	for index := range entries {
		entries[index] = SpendWindowEntryV1{Value: big.NewInt(1)}
	}
	if _, valid := SpendWindowTotalsV1(entries, 0, 1); valid {
		t.Fatal("oversized history accepted")
	}
	if _, valid := SpendWindowTotalsV1([]SpendWindowEntryV1{{Value: big.NewInt(-1)}}, 0, 1); valid {
		t.Fatal("negative value accepted")
	}
}

func TestSpendWindowTotalsV1SharedVectors(t *testing.T) {
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime caller unavailable")
	}
	data, err := os.ReadFile(filepath.Join(filepath.Dir(file), "../../../../packages/protocol/fixtures/spend-window-v1.json"))
	if err != nil {
		t.Fatal(err)
	}
	var fixture struct {
		Schema           string `json:"schema"`
		CalendarBoundary string `json:"calendarBoundary"`
		RollingBoundary  string `json:"rollingBoundary"`
		CaseCount        int    `json:"caseCount"`
		Cases            []struct {
			Name       string `json:"name"`
			Now        string `json:"now"`
			Window     string `json:"window"`
			EntryCount int    `json:"entryCount"`
			Entries    []struct {
				Value      string `json:"value"`
				ExecutedAt string `json:"executedAt"`
			} `json:"entries"`
			Daily   string `json:"daily"`
			Rolling string `json:"rolling"`
			Valid   bool   `json:"valid"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(data, &fixture); err != nil {
		t.Fatal(err)
	}
	if fixture.Schema != "SPEND_WINDOW_V1" || fixture.CalendarBoundary != "[dayStart, now]" || fixture.RollingBoundary != "(now-window, now]" || fixture.CaseCount != len(fixture.Cases) {
		t.Fatal("unexpected spend window fixture domain")
	}
	for _, vector := range fixture.Cases {
		t.Run(vector.Name, func(t *testing.T) {
			if vector.EntryCount != len(vector.Entries) {
				t.Fatal("entry count mismatch")
			}
			entries := make([]SpendWindowEntryV1, len(vector.Entries))
			for index, entry := range vector.Entries {
				entries[index] = SpendWindowEntryV1{Value: mathBig(entry.Value), ExecutedAt: spendWindowUint(t, entry.ExecutedAt)}
			}
			actual, valid := SpendWindowTotalsV1(entries, spendWindowUint(t, vector.Now), spendWindowUint(t, vector.Window))
			if valid != vector.Valid {
				t.Fatalf("validity mismatch: %v", valid)
			}
			if valid && (actual.DailySpend.Cmp(mathBig(vector.Daily)) != 0 || actual.RollingSpend.Cmp(mathBig(vector.Rolling)) != 0) {
				t.Fatalf("totals mismatch: %+v", actual)
			}
		})
	}
}

func spendWindowUint(t *testing.T, value string) uint64 {
	t.Helper()
	result, err := strconv.ParseUint(value, 10, 64)
	if err != nil {
		t.Fatal(err)
	}
	return result
}

File: packages/protocol/test/spend-window.test.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";
import { spendWindowTotalsV1 } from "../src/spend-window.js";

interface SpendWindowVector {
  name: string;
  now: string;
  window: string;
  entryCount: number;
  entries: { value: string; executedAt: string }[];
  daily: string;
  rolling: string;
  valid: boolean;
}

const fixture = JSON.parse(
  readFileSync(resolve(import.meta.dirname, "../fixtures/spend-window-v1.json"), "utf8"),
) as { schema: string; calendarBoundary: string; rollingBoundary: string; caseCount: number; cases: SpendWindowVector[] };

describe("SPEND_WINDOW_V1 shared vectors", () => {
  it("sums exact UTC calendar and sliding windows", () => {
    expect(fixture.schema).toBe("SPEND_WINDOW_V1");
    expect(fixture.calendarBoundary).toBe("[dayStart, now]");
    expect(fixture.rollingBoundary).toBe("(now-window, now]");
    expect(fixture.caseCount).toBe(fixture.cases.length);
    for (const vector of fixture.cases) {
      expect(vector.entryCount, vector.name).toBe(vector.entries.length);
      const actual = spendWindowTotalsV1(
        vector.entries.map((entry) => ({ value: BigInt(entry.value), executedAt: BigInt(entry.executedAt) })),
        BigInt(vector.now),
        BigInt(vector.window),
      );
      expect(actual, vector.name).toEqual(vector.valid
        ? { dailySpend: BigInt(vector.daily), rollingSpend: BigInt(vector.rolling) }
        : null);
    }
  });

  it("rejects invalid wire timestamps and window", () => {
    expect(spendWindowTotalsV1([], -1n, 1n)).toBeNull();
    expect(spendWindowTotalsV1([], 1n, 0n)).toBeNull();
    expect(spendWindowTotalsV1([{ value: -1n, executedAt: 1n }], 1n, 1n)).toBeNull();
  });

  it("caps private history work", () => {
    const oversized = Array.from({ length: 4_097 }, () => ({ value: 1n, executedAt: 0n }));
    expect(spendWindowTotalsV1(oversized, 0n, 1n)).toBeNull();
  });
});
